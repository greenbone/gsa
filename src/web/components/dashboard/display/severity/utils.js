/* SPDX-FileCopyrightText: 2025 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import FilterTerm from 'gmp/models/filter/filterterm';
import {isDefined} from 'gmp/utils/identity';
import {LOG_VALUE} from 'web/utils/severity';

/**
 * Converts filter values to filter terms based on severity range.
 *
 * @param {Object} params - The filter values.
 * @param {(string|number)} params.start - The start value of the filter range.
 * @param {(string|number)} params.end - The end value of the filter range.
 * @returns {FilterTerm[]} An array of FilterTerm objects representing the filter terms.
 */
export const filterValueToFilterTerms = ({start, end}) => {
  if (isDefined(end)) {
    const startTerm = FilterTerm.fromString(`severity>${start}`);
    const endTerm = FilterTerm.fromString(`severity<${end}`);
    return [startTerm, endTerm];
  }
  if (start > LOG_VALUE) {
    return [FilterTerm.fromString(`severity>${start}`)];
  }
  return [FilterTerm.fromString(`severity=${start}`)];
};
