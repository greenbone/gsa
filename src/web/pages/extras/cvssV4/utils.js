/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {CVSS40} from '@pandatix/js-cvss';
import {expectedMetricOptionsOrdered} from './cvssConfig';

/**
 * This function calculates the CVSS vector from a set of metrics.
 *
 * @param {Record<string,string>} cvssVectorObject - An object with key-value pairs of the metrics.
 * @returns {string} - The CVSS vector string.
 */

export const calculateVector = cvssVectorObject => {
  const result = expectedMetricOptionsOrdered
    .filter(metric => cvssVectorObject[metric[0]] !== undefined)
    .map(metric => `${metric[0]}:${cvssVectorObject[metric[0]]}`)
    .join('/');
  return result;
};

/**
 * This function processes a CVSS vector string and returns an object with the metrics.
 * Checks if the metric is a valid option.
 *
 * @param {string} vectorString - The CVSS vector string to process.
 * @returns {Record<string,string>} - An object with key-value pairs of the metrics.
 */

export const processVector = vectorString => {
  const vector = vectorString.replace(/^CVSS:4.0\//, '').split('/');
  const result = {};
  const expectedMetricMap = new Map(
    expectedMetricOptionsOrdered.map(metric => [metric[0], metric]),
  );

  vector.forEach(metricString => {
    const [key, value] = metricString.split(':');
    if (
      expectedMetricMap.has(key) &&
      value &&
      expectedMetricMap.get(key).includes(value)
    ) {
      result[key] = value;
    }
  });

  return result;
};

/**
 * This function removes the unused metrics from a CVSS vector.
 * The unused optional metrics are the ones with the value 'X'.
 *
 * @param {Record<string,string>} cvssVector - The CVSS vector to remove the unused metrics.
 * @returns {string} - The CVSS vector without the unused metrics.
 */

export const removeUnusedMetrics = cvssVector => {
  const vector = calculateVector(cvssVector).split('/');

  const validMetrics = vector.filter(metric => {
    const [, value] = metric.split(':');
    return value !== 'X';
  });

  const orderMap = new Map(
    expectedMetricOptionsOrdered.map((metric, index) => [metric[0], index]),
  );

  const getKey = metricString => metricString.split(':')[0];
  validMetrics.sort(
    (a, b) => orderMap.get(getKey(a)) - orderMap.get(getKey(b)),
  );

  return validMetrics.join('/');
};

/**
 * This function calculates the CVSS score from a CVSS vector.
 * @param {string} cvssVector - The CVSS vector with all the metrics to calculate the score.
 * @returns {number | undefined} - The CVSS score.
 *
 */

export const calculateScoreSafely = cvssVector => {
  try {
    return new CVSS40(cvssVector).Score();
  } catch {
    return undefined;
  }
};
