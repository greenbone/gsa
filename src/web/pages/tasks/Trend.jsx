/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import _ from 'gmp/locale';
import React from 'react';
import TrendDownIcon from 'web/components/icon/TrendDownIcon';
import TrendLessIcon from 'web/components/icon/TrendLessIcon';
import TrendMoreIcon from 'web/components/icon/TrendMoreIcon';
import TrendNoChangeIcon from 'web/components/icon/TrendNoChangeIcon';
import TrendUpIcon from 'web/components/icon/TrendUpIcon';
import PropTypes from 'web/utils/PropTypes';

const Trend = ({name}) => {
  let title;
  let IconComponent;

  if (name === 'up') {
    title = _('Severity increased');
    IconComponent = TrendUpIcon;
  } else if (name === 'down') {
    title = _('Severity decreased');
    IconComponent = TrendDownIcon;
  } else if (name === 'more') {
    title = _('Vulnerability count increased');
    IconComponent = TrendMoreIcon;
  } else if (name === 'less') {
    title = _('Vulnerability count decreased');
    IconComponent = TrendLessIcon;
  } else if (name === 'same') {
    title = _('Vulnerabilities did not change');
    IconComponent = TrendNoChangeIcon;
  } else {
    return <span />;
  }

  return <IconComponent alt={title} size="small" title={title} />;
};

Trend.propTypes = {
  name: PropTypes.string,
};

export default Trend;
