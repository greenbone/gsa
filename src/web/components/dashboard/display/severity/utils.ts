/* SPDX-FileCopyrightText: 2025 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import FilterTerm from 'gmp/models/filter/filterterm';
import {isDefined} from 'gmp/utils/identity';
import {LOG_VALUE} from 'web/utils/severity';

interface FilterValueToFilterTermsParams {
  start?: number;
  end?: number;
}

type FilterValueToFilterTermsReturnType = [FilterTerm?, FilterTerm?];

/**
 * Converts filter values to filter terms based on severity range.
 *
 * @param params - The filter values.
 * @param params.start - The start value of the filter range.
 * @param params.end - The end value of the filter range.
 * @returns An array of FilterTerm objects representing the filter terms.
 */
export const filterValueToFilterTerms = ({
  start,
  end,
}: FilterValueToFilterTermsParams): FilterValueToFilterTermsReturnType => {
  if (!isDefined(start)) {
    return [];
  }
  if (isDefined(end)) {
    const startTerm = FilterTerm.fromString(`severity>${start}`);
    const endTerm = FilterTerm.fromString(`severity<${end}`);
    return [startTerm, endTerm];
  }
  if (start > LOG_VALUE) {
    return [FilterTerm.fromString(`severity>${start}`)];
  }
  return [FilterTerm.fromString(`severity=${start}`)];
};
