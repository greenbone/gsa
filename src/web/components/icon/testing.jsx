/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */


import {test, expect, testing} from '@gsa/testing';

import {render, fireEvent, act} from 'web/utils/testing';
import Theme from 'web/utils/theme';

import {ICON_SIZE_SMALL_PIXELS} from 'web/hooks/useIconSize';

export const testIcon = Icon => {
  test('should render with default width and height', () => {
    const {element} = render(<Icon />);

    expect(element).toHaveStyleRule('width', ICON_SIZE_SMALL_PIXELS);
    expect(element).toHaveStyleRule('height', ICON_SIZE_SMALL_PIXELS);
  });

  test('should handle click', () => {
    const handler = testing.fn();
    const {element} = render(<Icon onClick={handler} value="1" />);

    fireEvent.click(element);

    expect(handler).toHaveBeenCalledWith('1');
  });
  test('should change appearance when disabled', () => {
    const {element} = render(<Icon disabled={true} />);

    expect(element).toHaveStyleRule('fill', Theme.inputBorderGray, {
      modifier: 'svg path',
    });
  });

  test('should change appearance when inactive', () => {
    const {element} = render(<Icon active={false} />);

    expect(element).toHaveStyleRule('fill', Theme.inputBorderGray, {
      modifier: '& svg path',
    });
  });

  test('should not call clickhandler when disabled', () => {
    const handler = testing.fn();
    const {element} = render(<Icon disabled={true} onClick={handler} />);

    fireEvent.click(element);

    expect(handler).not.toHaveBeenCalled();
  });

  test('should render loading state', async () => {
    let resolve;
    const promise = new Promise((res, rej) => {
      resolve = res;
    });

    const handler = testing.fn().mockReturnValue(promise);

    const {element} = render(
      <Icon title="foo" loadingTitle="bar" onClick={handler} />,
    );

    const svgElement = element.querySelector('svg');

    expect(element).toHaveAttribute('title', 'foo');
    expect(svgElement).toHaveAttribute('title', 'foo');

    fireEvent.click(element);

    expect(element).toHaveStyleRule('display', 'none', {
      modifier: '& svg path',
    });
    expect(element).toHaveStyleRule(
      'background',
      'url(/img/loading.gif) center center no-repeat',
      {
        modifier: '& svg',
      },
    );

    expect(element).toHaveAttribute('title', 'bar');
    expect(svgElement).toHaveAttribute('title', 'bar');

    await act(async () => {
      resolve();
    });

    expect(element).toHaveAttribute('title', 'foo');
    expect(svgElement).toHaveAttribute('title', 'foo');
  });
};

// vim: set ts=2 sw=2 tw=80:
