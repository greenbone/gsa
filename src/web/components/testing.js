/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {isDefined} from 'gmp/utils/identity';

import {
  userEvent,
  act,
  fireEvent,
  queryAllByRole,
  queryByRole,
  getByRole,
  getAllByTestId,
  getByTestId,
} from 'web/utils/testing';

export const getElementOrDocument = element =>
  isDefined(element) ? element : document.body;

/**
 * Get all the radio inputs
 */
export const getRadioInputs = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-Radio-radio');
};

/**
 * Get all the items of a Select component
 */
export const getSelectItemElements = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll("[role='option']");
};

/**
 * Get the items that are selectable of a Select component
 */
export const getSelectItemElementsForSelect = async element => {
  element = isDefined(element) ? element : getSelectElement();
  await openSelectElement(element);
  const selectItemsId = element.getAttribute('aria-controls');
  const itemsContainer = document.body.querySelector('#' + selectItemsId);
  return getSelectItemElements(itemsContainer);
};

/**
 * Get the input box of a Select component
 */
export const getSelectElement = element => {
  element = getElementOrDocument(element);
  const select = queryByRole(element, 'searchbox');
  if (select) {
    return select;
  }
  return getByRole(element, 'textbox');
};

/**
 * Get all select components
 */
export const getSelectElements = element => {
  element = getElementOrDocument(element);
  return queryAllByRole(element, 'searchbox');
};

/**
 * Open a select element (MultiSelect, Select, etc.)
 */
export const openSelectElement = async select => {
  await act(async () => {
    select = select || getSelectElement();
    await userEvent.click(select);
  });
};

/**
 * Click on an element/item/node
 */
export const clickElement = async element => {
  await act(async () => {
    await userEvent.click(element);
  });
};

/**
 * Get all multi select (root) elements
 */
export const getMultiSelectElements = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-root');
};

/**
 * Get all multi select (root) element
 */
export const getMultiSelectElement = element => {
  element = getElementOrDocument(element);
  return element.querySelector('.mantine-MultiSelect-root');
};

/**
 * Open a MultiSelect component for displaying item selection
 */
export const openMultiSelectElement = async element => {
  element = isDefined(element) ? element : getMultiSelectElement();
  const input = getSelectElement(element);
  return await clickElement(input);
};

/**
 * Get current selected items of a MultiSelect component
 */
export const getSelectedItems = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-value');
};

/**
 * Get all the selectable items of a MultiSelect component
 */
export const getSelectItemElementsForMultiSelect = async element => {
  element = isDefined(element) ? element : getMultiSelectElement();
  await openMultiSelectElement(element);
  const controlElement = element.querySelector('.mantine-MultiSelect-wrapper');
  const selectItemsId = controlElement.getAttribute('aria-controls');
  const itemsContainer = document.body.querySelector(
    '#' + selectItemsId + '-items',
  );
  return getSelectItemElements(itemsContainer);
};

/**
 * Change the value of an input element like Select or TestField component
 */
export const changeInputValue = (element, value) => {
  fireEvent.change(element, {target: {value}});
};

/**
 * Change the input of a Select component
 */
export const changeSelectInput = (value, input) => {
  if (!isDefined(input)) {
    input = getSelectElement();
  }
  changeInputValue(input, value);
};

/**
 * Query if a dialog is present
 */
export const queryDialog = element => {
  element = getElementOrDocument(element);
  return queryByRole(element, 'dialog');
};

/**
 * Get a dialog
 */
export const getDialog = element => {
  element = getElementOrDocument(element);
  return getByRole(element, 'dialog');
};

/**
 * Get the dialog content
 */
export const getDialogContent = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-body');
};

/**
 * Get the dialog title
 */
export const getDialogTitle = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-title');
};

/**
 * Get the save button of a dialog
 */
export const getDialogSaveButton = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-save-button');
};

/**
 * Get the close button of a dialog (in the footer)
 */
export const getDialogCloseButton = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-close-button');
};

/**
 * Close a dialog
 */
export const closeDialog = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  const closeButton = dialog.querySelector('.mantine-CloseButton-root');
  fireEvent.click(closeButton);
};

/**
 * Get the element containing the powerfilter
 */
export const getPowerFilter = element => {
  element = getElementOrDocument(element);
  return element.querySelector('.powerfilter');
};

/**
 * Get text inputs
 */
export const getTextInputs = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-TextInput-input');
};

/**
 * Get the table element
 */
export const getTable = element => {
  element = getElementOrDocument(element);
  return element.querySelector('table');
};

/**
 * Get the table body element
 */
export const getTableBody = element => {
  element = getElementOrDocument(element);
  return element.querySelector('tbody');
};

/**
 * Get the table footer element
 */
export const getTableFooter = element => {
  element = getElementOrDocument(element);
  return element.querySelector('tfoot');
};

/**
 * Get the table header element
 */
export const getTableHeader = element => {
  element = getElementOrDocument(element);
  return element.querySelector('thead');
};

/**
 * Get action items
 */
export const getActionItems = element => {
  element = getElementOrDocument(element);
  return getAllByTestId(element, 'svg-icon');
};

/**
 * Get the bulk action items of a page
 */
export const getBulkActionItems = element => {
  const tableFooter = getTableFooter(element);
  return getActionItems(tableFooter);
};

/**
 * Get the check boxes
 */
export const getCheckBoxes = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-Checkbox-input');
};
