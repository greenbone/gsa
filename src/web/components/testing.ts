/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {PointerEventsCheckLevel} from '@testing-library/user-event';
import {isDefined} from 'gmp/utils/identity';
import {expect, Mock} from 'vitest';
import {
  userEvent,
  fireEvent,
  queryByRole,
  getByRole,
  getByTestId,
  queryAllByTestId,
  queryAllByRole,
  screen,
} from 'web/utils/Testing';

/**
 * Returns the provided element if it is defined, otherwise returns the document body.
 *
 * @param element - The element to check.
 * @returns The provided element if defined, otherwise the document body.
 */
export const getElementOrReturnDocument = (element?: HTMLElement) =>
  isDefined(element) ? element : document.body;

/**
 * Get all the radio inputs within the specified element.
 *
 * @param element - The optional parent element or document to search within.
 * @returns A NodeList of radio input elements.
 */
export const getRadioInputs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-Radio-radio');
};

/**
 * Get all the items of a Select component
 *
 * @param element - The optional root element or document to search within.
 * @returns A NodeList of select item elements
 */
export const getSelectItemElements = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll("[role='option']");
};

/**
 * Retrieves the selectable item elements for a given select element.
 *
 * @param element - The optional select element to retrieve items for. If not provided, a default select element will be used.
 * @returns A promise that resolves to a NodeList of the selectable item elements.
 */
export const getSelectItemElementsForSelect = async (
  element?: HTMLSelectElement,
) => {
  element = isDefined(element) ? element : getSelectElement();
  await openSelectElement(element);
  const selectItemsId = element.getAttribute('aria-controls');
  const itemsContainer =
    document.body.querySelector<HTMLElement>('#' + selectItemsId) || undefined;
  return getSelectItemElements(itemsContainer);
};

/**
 * Get the input box of a Select component
 *
 * This function first attempts to find an element with the test ID 'form-select'.
 * If no such element is found, it falls back to finding an element with the role 'textbox'.
 *
 * @param element - The optional element to search within.
 * @returns The found select element
 */
export const getSelectElement = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  const select = getByTestId(element, 'form-select');
  if (select) {
    return select as HTMLSelectElement;
  }
  return getByRole(element, 'textbox') as HTMLSelectElement;
};

/**
 * Get all select components
 *
 * @param element - The root element to query within. If not provided, the document will be used.
 * @returns A list of select elements
 */
export const queryAllSelectElements = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return queryAllByTestId(element, 'form-select');
};

/**
 * Open a select element (MultiSelect, Select, etc.)
 *
 * @param select - An optional select element to open. If not provided, it defaults to the result of getSelectElement().
 * @returns A promise that resolves when the select element has been opened.
 */
export const openSelectElement = async (select?: HTMLSelectElement) => {
  select = select || getSelectElement();
  await clickElement(select);
};

/**
 * Clicks on the given element/item/node using userEvent.
 *
 * @param element - The element to be clicked.
 * @returns A promise that resolves when the click action is complete.
 */
export const clickElement = async (element: Element) => {
  await userEvent.click(element, {
    pointerEventsCheck: PointerEventsCheckLevel.Never,
  });
};

/**
 * Get all multi select (root) elements
 *
 * @param element - An optional element to search within. If not provided, the document will be used.
 * @returns A NodeList of multi select elements.
 */
export const getMultiSelectElements = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-inputField');
};

/**
 * Retrieves the multi select (root) element from the given element.
 *
 * @returns The found multi select element.
 */
export const getMultiSelectElement = () => {
  return screen.getByRole('textbox');
};

/**
 * Open a MultiSelect component for displaying item selection
 *
 * @param element - The multi-select element to open. If not defined, it will use a default multi-select element.
 * @returns A promise that resolves when the input element has been clicked.
 */
export const openMultiSelectElement = async (element?: HTMLElement) => {
  element = isDefined(element) ? element : getMultiSelectElement();
  const input = getSelectElement(element);
  return await clickElement(input);
};

/**
 * Get current selected items of a MultiSelect component
 *
 * @param element - The document object to query.
 * @returns A NodeList of selected items.
 */
export const getSelectedItems = (element: ParentNode) => {
  return element.querySelectorAll<HTMLElement>('.mantine-MultiSelect-pill');
};

/**
 * Get all the selectable items of a MultiSelect component from screen.
 *
 * @returns An array of select item elements.
 */
export const getSelectItemElementsForMultiSelect = () => {
  return screen.getAllByRole('option');
};

/**
 * Changes the value of an input element like Select or TestField component and triggers the blur event.
 *
 * @param element - The input element whose value is to be changed.
 * @param value - The new value to be set for the input element.
 */
export const changeInputValue = (
  element: Document | Element | Window | Node,
  value: string,
) => {
  fireEvent.change(element, {target: {value}});
  fireEvent.blur(element);
};

/**
 * Changes the value of a select input element.
 *
 * @param value - The new value to set for the select input.
 * @param input - The select input element to change. If not provided, a default select element will be used.
 */
export const changeSelectInput = (value: string, input?: HTMLSelectElement) => {
  if (!isDefined(input)) {
    input = getSelectElement();
  }
  changeInputValue(input, value);
};

/**
 * Queries all dialog elements within the specified element or document.
 *
 * @param element - The element or document to query within. If not provided, the document will be used.
 * @returns A list of dialogs.
 */
export const queryDialogs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return queryAllByRole(element, 'dialog');
};

/**
 * Queries if a dialog is present
 *
 * @param element - The element to query within. If not provided, the document will be used.
 * @returns The dialog element if found, otherwise null.
 */
export const queryDialog = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return queryByRole(element, 'dialog');
};

/**
 * Retrieves the dialog
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The dialog element found within the specified element or document.
 */
export const getDialog = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return getByRole(element, 'dialog');
};

/**
 * Retrieves the content of a dialog.
 *
 * @param dialog - The dialog element. If not defined, a default dialog is retrieved.
 * @returns The content element of the dialog, or null if not found.
 */
export const queryDialogContent = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-body');
};

/**
 * Retrieves the title element from a dialog.
 *
 * @param dialog - The dialog element. If not provided, a default dialog is retrieved using the getDialog function.
 * @returns The title element of the dialog, or null if not found.
 */
export const queryDialogTitle = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-title');
};

/**
 * Retrieves the save button of a dialog (in the footer)
 *
 * @param dialog - The dialog element. If not provided, the default dialog will be used.
 * @returns The save button element within the dialog.
 */
export const getDialogSaveButton = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-save-button');
};

/**
 * Retrieves the close button of a dialog (in the footer).
 *
 * @param dialog - The dialog element. If not provided, the default dialog will be used.
 * @returns The close button element within the dialog.
 */
export const getDialogCloseButton = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-close-button');
};

/**
 * Closes a dialog by clicking its close (X) button.
 *
 * @param dialog - The dialog element to close. If not provided, it defaults to the result of getDialog().
 */
export const closeDialog = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  const closeButton = dialog.querySelector(
    '.mantine-CloseButton-root',
  ) as HTMLElement;
  fireEvent.click(closeButton);
};

/**
 * Get the element containing the powerfilter
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The powerfilter element, or null if not found.
 */
export const queryPowerFilter = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('.powerfilter');
};

/**
 * Get text inputs
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns A NodeList of text input elements
 */
export const queryTextInputs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-TextInput-input');
};

/**
 * Queries the table element
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The first table element found within the given element, or null if no table is found.
 */
export const queryTable = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('table');
};

/**
 * Queries the table body element
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The <tbody> element if found, otherwise null.
 */
export const queryTableBody = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('tbody');
};

/**
 * Queries the table footer element.
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The `<tfoot>` element if found, otherwise `null`.
 */
export const queryTableFooter = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('tfoot');
};

/**
 * Queries the table header element.
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The <thead> element of the table, or null if not found.
 */
export const queryTableHeader = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('thead');
};

/**
 * Queries and returns all checkbox input elements
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns A NodeList of all checkbox input elements found within the specified element.
 */
export const queryCheckBoxes = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-Checkbox-input');
};

/**
 * Queries and returns all file input elements
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns A NodeList of file input elements
 */
export const queryFileInputs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-FileInput-input');
};

export const testBulkTrashcanDialog = (
  _screen: unknown,
  dialogAction: Mock,
) => {
  const dialog = screen.getByRole('dialog');
  expect(dialog).toBeVisible();

  const moveToTrashcanButton = screen.getByText('Move to Trashcan');
  fireEvent.click(moveToTrashcanButton);

  expect(dialogAction).toHaveBeenCalled();
};

export const testBulkDeleteDialog = (_screen: unknown, dialogAction: Mock) => {
  const dialog = screen.getByTestId('confirmation-dialog');
  expect(dialog).toBeVisible();

  const title = screen.getByText('Confirm Deletion');

  expect(title).toBeVisible();

  const deleteButton = screen.getByText('Delete');
  fireEvent.click(deleteButton);

  expect(dialogAction).toHaveBeenCalled();
};
