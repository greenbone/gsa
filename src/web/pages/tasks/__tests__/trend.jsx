/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {describe, test, expect} from '@gsa/testing';

import {render} from 'web/utils/testing';

import Trend from '../trend';

describe('Task Trend tests', () => {
  test('should render', () => {
    const {element} = render(<Trend name="up" />);

    expect(element).toBeInTheDocument();
  });

  test('should render trend up icon', () => {
    const {element} = render(<Trend name="up" />);

    expect(element).toHaveAttribute('title', 'Severity increased');
  });

  test('should render trend down icon', () => {
    const {element} = render(<Trend name="down" />);

    expect(element).toHaveAttribute('title', 'Severity decreased');
  });

  test('should render trend less icon', () => {
    const {element} = render(<Trend name="less" />);

    expect(element).toHaveAttribute('title', 'Vulnerability count decreased');
  });

  test('should render trend more icon', () => {
    const {element} = render(<Trend name="more" />);

    expect(element).toHaveAttribute('title', 'Vulnerability count increased');
  });

  test('should render trend no change icon', () => {
    const {element} = render(<Trend name="same" />);

    expect(element).toHaveAttribute('title', 'Vulnerabilities did not change');
  });
});
