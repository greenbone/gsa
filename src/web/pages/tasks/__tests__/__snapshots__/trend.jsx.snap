// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task Trend tests > should render 1`] = `
.c0 {
  width: 16px;
  height: 16px;
  line-height: 16px;
}

.c0 * {
  height: inherit;
  width: inherit;
}

<span
  alt="Severity increased"
  class="c0"
  data-testid="svg-icon"
  title="Severity increased"
>
  <svg
    height="1em"
    title="Severity increased"
    viewBox="0 0 67.733 67.733"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      transform="translate(0 -229.27)"
    >
      <g
        transform="matrix(2.0878 0 0 2.2337 41.673 -337.41)"
      >
        <path
          class="gui_icon_class"
          d="m-16.862 283.13 7.3086-16.227 5.62 2.1758 4.6557-10.506-4.6989-1.8192 7.9479-3.0771 3.2921 7.4288-4.6678-1.8072-5.4317 12.257-5.62-2.1758-6.557 14.603z"
          fill-rule="evenodd"
        />
      </g>
    </g>
  </svg>
</span>
`;
